File: packages/ruby/bin/benchmark.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'optparse'
require 'time'

$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))
require 'html_to_markdown'

def json_escape(value)
  value.to_s.gsub(/["\\\n\r]/) do |char|
    case char
    when '"', '\\'
      "\\#{char}"
    when "\n"
      '\\n'
    when "\r"
      '\\r'
    end
  end
end

options = {
  iterations: 50,
  format: 'html',
  scenario: 'convert-default',
  visitor: nil
}

OptionParser.new do |parser|
  parser.banner = 'ruby benchmark.rb --file path/to/fixture.html [--iterations 200]'

  parser.on('--file FILE', 'HTML fixture to convert repeatedly') do |file|
    options[:file] = file
  end

  parser.on('--iterations N', Integer, 'Number of conversion iterations (default: 50)') do |n|
    options[:iterations] = n.positive? ? n : 1
  end

  parser.on('--scenario SCENARIO', 'Scenario to benchmark') do |scenario|
    options[:scenario] = scenario
  end

  parser.on('--format FORMAT', 'Fixture format (html or hocr)') do |format|
    options[:format] = format.downcase
  end

  parser.on('--visitor VISITOR', 'Visitor type (noop, simple, custom, complex)') do |visitor|
    options[:visitor] = visitor if %w[noop simple custom complex].include?(visitor)
  end
end.parse!

fixture = options.fetch(:file) do
  warn 'Missing --file parameter'
  exit 1
end

unless File.exist?(fixture)
  warn "Fixture not found: #{fixture}"
  exit 1
end

unless %w[html hocr].include?(options[:format])
  warn "Unsupported format: #{options[:format]}"
  exit 1
end

supported_scenarios = %w[
  convert-default
  convert-options
  inline-images-default
  inline-images-options
  metadata-default
  metadata-options
]
unless supported_scenarios.include?(options[:scenario])
  warn "Unsupported scenario: #{options[:scenario]}"
  exit 1
end

# Visitor factory functions
def create_noop_visitor
  {
    visit_text: proc { |_ctx, _text| 'continue' },
    visit_heading: proc { |_ctx, _level, _text, _id| 'continue' },
    visit_paragraph: proc { |_ctx, _text| 'continue' },
    visit_link: proc { |_ctx, _href, _text, _title| 'continue' },
    visit_image: proc { |_ctx, _src, _alt, _title| 'continue' },
    visit_strong: proc { |_ctx, _text| 'continue' },
    visit_em: proc { |_ctx, _text| 'continue' },
    visit_code: proc { |_ctx, _text| 'continue' },
    visit_br: proc { |_ctx| 'continue' }
  }
end

def create_simple_visitor
  {
    text_count: 0,
    link_count: 0,
    image_count: 0,
    visit_text: proc { |_ctx, _text| 'continue' },
    visit_heading: proc { |_ctx, _level, _text, _id| 'continue' },
    visit_paragraph: proc { |_ctx, _text| 'continue' },
    visit_link: proc { |_ctx, _href, _text, _title| 'continue' },
    visit_image: proc { |_ctx, _src, _alt, _title| 'continue' },
    visit_strong: proc { |_ctx, _text| 'continue' },
    visit_em: proc { |_ctx, _text| 'continue' },
    visit_code: proc { |_ctx, _text| 'continue' },
    visit_br: proc { |_ctx| 'continue' }
  }
end

def create_custom_visitor
  {
    visit_text: proc { |_ctx, _text| 'continue' },
    visit_heading: proc { |_ctx, _level, _text, _id| 'continue' },
    visit_paragraph: proc { |_ctx, _text| 'continue' },
    visit_link: proc { |_ctx, href, text, _title| ['custom', "LINK[#{text}](#{href})"] },
    visit_image: proc { |_ctx, src, alt, _title| ['custom', "![#{alt}](#{src})"] },
    visit_strong: proc { |_ctx, _text| 'continue' },
    visit_em: proc { |_ctx, _text| 'continue' },
    visit_code: proc { |_ctx, _text| 'continue' },
    visit_br: proc { |_ctx| 'continue' }
  }
end

def create_complex_visitor
  {
    texts: 0,
    links: 0,
    images: 0,
    headings: 0,
    visit_text: proc { |_ctx, _text| 'continue' },
    visit_heading: proc { |_ctx, _level, _text, _id| 'continue' },
    visit_paragraph: proc { |_ctx, _text| 'continue' },
    visit_link: proc { |_ctx, href, text, _title| ['custom', "[#{text}](#{href})"] },
    visit_image: proc { |_ctx, _src, _alt, _title| 'skip' },
    visit_strong: proc { |_ctx, _text| 'continue' },
    visit_em: proc { |_ctx, _text| 'continue' },
    visit_code: proc { |_ctx, _text| 'continue' },
    visit_br: proc { |_ctx| 'continue' }
  }
end

html = File.binread(fixture)
html.force_encoding(Encoding::UTF_8)
html.freeze
iterations = options[:iterations]
conversion_options = options[:format] == 'hocr' ? { hocr_spatial_tables: false } : {}
options_handle = if %w[convert-options inline-images-options metadata-options].include?(options[:scenario])
                   HtmlToMarkdown.options(conversion_options)
                 end

# Create visitor if specified
visitor = nil
if options[:visitor]
  visitor_creators = {
    'noop' => method(:create_noop_visitor),
    'simple' => method(:create_simple_visitor),
    'custom' => method(:create_custom_visitor),
    'complex' => method(:create_complex_visitor)
  }
  creator = visitor_creators[options[:visitor]]
  visitor = creator.call if creator
end

SCENARIO_RUNNERS = {
  'convert-default' => ->(html, _options, _handle, _visitor) { HtmlToMarkdown.convert(html) },
  'convert-options' => lambda do |html, _options, handle, _visitor|
    raise ArgumentError, 'options handle required' unless handle

    HtmlToMarkdown.convert_with_options(html, handle)
  end,
  'inline-images-default' => lambda { |html, _options, _handle, _visitor|
    HtmlToMarkdown.convert_with_inline_images(html, nil, nil)
  },
  'inline-images-options' => lambda do |html, _options, handle, _visitor|
    raise ArgumentError, 'options handle required' unless handle

    HtmlToMarkdown.convert_with_inline_images_handle(html, handle, nil)
  end,
  'metadata-default' => ->(html, _options, _handle, _visitor) { HtmlToMarkdown.convert_with_metadata(html, nil, nil) },
  'metadata-options' => lambda do |html, _options, handle, _visitor|
    raise ArgumentError, 'options handle required' unless handle

    HtmlToMarkdown.convert_with_metadata_handle(html, handle, nil)
  end
}.freeze

def run_scenario(html, scenario, options, handle, visitor = nil)
  if visitor
    HtmlToMarkdown.convert_with_visitor(html, nil, visitor)
  else
    runner = SCENARIO_RUNNERS.fetch(scenario) { raise ArgumentError, "Unsupported scenario: #{scenario}" }
    runner.call(html, options, handle, visitor)
  end
end

run_scenario(html, options[:scenario], conversion_options, options_handle, visitor)

profile_output = ENV.fetch('HTML_TO_MARKDOWN_PROFILE_OUTPUT', nil)
if profile_output && HtmlToMarkdown.respond_to?(:start_profiling)
  freq = Integer(ENV.fetch('HTML_TO_MARKDOWN_PROFILE_FREQUENCY', '1000'), 10)
  HtmlToMarkdown.start_profiling(profile_output, freq)
end

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
iterations.times { run_scenario(html, options[:scenario], conversion_options, options_handle, visitor) }
elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start

HtmlToMarkdown.stop_profiling if profile_output && HtmlToMarkdown.respond_to?(:stop_profiling)

payload_size_bytes = html.bytesize
bytes_processed = payload_size_bytes * iterations
ops_per_sec = iterations / elapsed
mb_per_sec = (bytes_processed.to_f / (1024 * 1024)) / elapsed

payload = %({
  "language":"ruby",
  "fixture":"#{json_escape(File.basename(fixture))}",
  "fixture_path":"#{json_escape(fixture)}",
  "scenario":"#{json_escape(options[:scenario])}",
  "iterations":#{iterations},
  "elapsed_seconds":#{format('%.8f', elapsed)},
  "ops_per_sec":#{format('%.4f', ops_per_sec)},
  "mb_per_sec":#{format('%.4f', mb_per_sec)},
  "bytes_processed":#{bytes_processed},
  "payload_size_bytes":#{payload_size_bytes}
})

puts payload.strip

File: packages/python/bin/benchmark.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import sys
import time
from pathlib import Path
from typing import TYPE_CHECKING

from html_to_markdown import (
    ConversionOptions,
    InlineImageConfig,
    MetadataConfig,
    OptionsHandle,
    convert,
    convert_with_handle,
    convert_with_inline_images,
    convert_with_inline_images_handle,
    convert_with_metadata,
    convert_with_metadata_handle,
    convert_with_visitor,
    create_options_handle,
    start_profiling,
    stop_profiling,
)

if TYPE_CHECKING:
    from collections.abc import Callable


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Benchmark html-to-markdown Python bindings")
    parser.add_argument("--file", required=True, help="Path to the HTML/HOCR fixture")
    parser.add_argument("--iterations", type=int, default=50, help="Number of iterations")
    parser.add_argument(
        "--scenario",
        default="convert-default",
        choices=(
            "convert-default",
            "convert-options",
            "inline-images-default",
            "inline-images-options",
            "metadata-default",
            "metadata-options",
        ),
        help="Benchmark scenario to run",
    )
    parser.add_argument(
        "--format",
        choices=("html", "hocr"),
        default="html",
        help="Fixture format to guide option defaults",
    )
    parser.add_argument(
        "--visitor",
        choices=("noop", "simple", "custom", "complex"),
        default=None,
        help="Visitor type to use during conversion",
    )
    return parser.parse_args()


def build_options(fixture_format: str) -> ConversionOptions:
    if fixture_format == "hocr":
        return ConversionOptions(hocr_spatial_tables=False)
    return ConversionOptions()


class NoopVisitor:
    def visit_text(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_heading(self, _ctx: object, _level: int, _text: str, _element_id: str) -> str:
        return "continue"

    def visit_paragraph(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_link(self, _ctx: object, _href: str, _text: str, _title: str) -> str:
        return "continue"

    def visit_image(self, _ctx: object, _src: str, _alt: str, _title: str) -> str:
        return "continue"

    def visit_strong(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_em(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_code(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_br(self, _ctx: object) -> str:
        return "continue"


def create_noop_visitor() -> object:
    """Create a no-op visitor that returns Continue for all methods."""
    return NoopVisitor()


class SimpleVisitor:
    def __init__(self) -> None:
        self.text_count = 0
        self.link_count = 0
        self.image_count = 0

    def visit_text(self, _ctx: object, _text: str) -> str:
        self.text_count += 1
        return "continue"

    def visit_heading(self, _ctx: object, _level: int, _text: str, _element_id: str) -> str:
        return "continue"

    def visit_paragraph(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_link(self, _ctx: object, _href: str, _text: str, _title: str) -> str:
        self.link_count += 1
        return "continue"

    def visit_image(self, _ctx: object, _src: str, _alt: str, _title: str) -> str:
        self.image_count += 1
        return "continue"

    def visit_strong(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_em(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_code(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_br(self, _ctx: object) -> str:
        return "continue"


def create_simple_visitor() -> object:
    """Create a simple visitor with basic callbacks."""
    return SimpleVisitor()


class CustomVisitor:
    def visit_text(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_heading(self, _ctx: object, _level: int, _text: str, _element_id: str) -> str:
        return "continue"

    def visit_paragraph(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_link(self, _ctx: object, _href: str, text: str, _title: str) -> tuple[str, str]:
        return ("custom", f"LINK[{text}]({_href})")

    def visit_image(self, _ctx: object, src: str, alt: str, _title: str) -> tuple[str, str]:
        return ("custom", f"![{alt}]({src})")

    def visit_strong(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_em(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_code(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_br(self, _ctx: object) -> str:
        return "continue"


def create_custom_visitor() -> object:
    """Create a custom output visitor that modifies conversion."""
    return CustomVisitor()


class ComplexVisitor:
    def __init__(self) -> None:
        self.stats: dict[str, int] = {"texts": 0, "links": 0, "images": 0, "headings": 0}

    def visit_text(self, _ctx: object, _text: str) -> str:
        self.stats["texts"] += 1
        return "continue"

    def visit_heading(self, _ctx: object, _level: int, _text: str, _element_id: str) -> str:
        self.stats["headings"] += 1
        return "continue"

    def visit_paragraph(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_link(self, _ctx: object, _href: str, text: str, _title: str) -> tuple[str, str]:
        self.stats["links"] += 1
        return ("custom", f"[{text}]({_href})")

    def visit_image(self, _ctx: object, _src: str, _alt: str, _title: str) -> str:
        self.stats["images"] += 1
        return "skip"

    def visit_strong(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_em(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_code(self, _ctx: object, _text: str) -> str:
        return "continue"

    def visit_br(self, _ctx: object) -> str:
        return "continue"


def create_complex_visitor() -> object:
    """Create a complex visitor with multiple callbacks."""
    return ComplexVisitor()


def require_handle(handle: OptionsHandle | None, scenario: str) -> OptionsHandle:
    if handle is None:
        raise SystemExit(f"Options handle required for {scenario} scenario")
    return handle


def require_metadata(metadata_config: MetadataConfig | None, scenario: str) -> MetadataConfig:
    if metadata_config is None:
        raise SystemExit(f"Metadata config required for {scenario} scenario")
    return metadata_config


def run_scenario(
    html: str,
    scenario: str,
    handle: OptionsHandle | None,
    metadata_config: MetadataConfig | None,
    visitor: object | None = None,
) -> None:
    if visitor is not None:
        # When visitor is provided, use convert_with_visitor for all scenarios
        convert_with_visitor(html, None, visitor)  # type: ignore[arg-type]
        return

    handlers: dict[str, Callable[[], object]] = {
        "convert-default": lambda: convert(html),
        "convert-options": lambda: convert_with_handle(html, require_handle(handle, "convert-options")),
        "inline-images-default": lambda: convert_with_inline_images(html, None, None, InlineImageConfig()),
        "inline-images-options": lambda: convert_with_inline_images_handle(
            html, require_handle(handle, "inline-images-options"), InlineImageConfig()
        ),
        "metadata-default": lambda: convert_with_metadata(
            html, None, None, require_metadata(metadata_config, "metadata-default")
        ),
        "metadata-options": lambda: convert_with_metadata_handle(
            html,
            require_handle(handle, "metadata-options"),
            require_metadata(metadata_config, "metadata-options"),
        ),
    }
    handler = handlers.get(scenario)
    if handler is None:
        raise SystemExit(f"Unsupported scenario: {scenario}")
    handler()


def main() -> None:
    args = parse_args()
    iterations = max(1, args.iterations)
    fixture = Path(args.file)

    if not fixture.exists():
        raise SystemExit(f"Fixture not found: {fixture}")

    html = fixture.read_text(encoding="utf-8")
    options = build_options(args.format)
    handle = (
        create_options_handle(options)
        if args.scenario in {"convert-options", "inline-images-options", "metadata-options"}
        else None
    )
    metadata_config = MetadataConfig() if "metadata" in args.scenario else None

    # Create visitor if specified
    visitor = None
    if args.visitor:
        visitor_creators = {
            "noop": create_noop_visitor,
            "simple": create_simple_visitor,
            "custom": create_custom_visitor,
            "complex": create_complex_visitor,
        }
        creator = visitor_creators.get(args.visitor)
        if creator:
            visitor = creator()

    run_scenario(html, args.scenario, handle, metadata_config, visitor)  # Warmup

    profile_output = os.getenv("HTML_TO_MARKDOWN_PROFILE_OUTPUT")
    profile_frequency = os.getenv("HTML_TO_MARKDOWN_PROFILE_FREQUENCY")
    if profile_output:
        freq = int(profile_frequency) if profile_frequency and profile_frequency.isdigit() else 1000
        start_profiling(profile_output, freq)

    start = time.perf_counter()
    for _ in range(iterations):
        run_scenario(html, args.scenario, handle, metadata_config, visitor)
    elapsed = time.perf_counter() - start

    if profile_output:
        stop_profiling()

    bytes_processed = len(html.encode("utf-8")) * iterations
    ops_per_sec = iterations / elapsed
    mb_per_sec = (bytes_processed / (1024 * 1024)) / elapsed

    result = {
        "language": "python",
        "fixture": fixture.name,
        "fixture_path": str(fixture),
        "scenario": args.scenario,
        "iterations": iterations,
        "elapsed_seconds": elapsed,
        "ops_per_sec": ops_per_sec,
        "mb_per_sec": mb_per_sec,
        "bytes_processed": bytes_processed,
    }

    sys.stdout.write(json.dumps(result) + "\n")


if __name__ == "__main__":
    main()
